File: python-sdk/captcha_background_sdk/font_glyph_images.py
from __future__ import annotations

from pathlib import Path
from typing import Iterable, List

from PIL import Image

from .types import (
    FontGlyph,
    FontGlyphImageExportResult,
    FontGlyphImageItem,
    GlyphRenderMode,
    GlyphRenderModeLike,
    normalize_glyph_render_mode,
)


def _render_rgba_pixel(pixel: tuple[int, int, int, int], render_mode: GlyphRenderMode) -> tuple[int, int, int, int]:
    r, g, b, a = pixel
    if render_mode == GlyphRenderMode.ORIGINAL:
        return (r, g, b, a)
    is_fg = a > 0
    if render_mode == GlyphRenderMode.BLACK_ON_TRANSPARENT:
        return (0, 0, 0, 255) if is_fg else (0, 0, 0, 0)
    if render_mode == GlyphRenderMode.BLACK_ON_WHITE:
        return (0, 0, 0, 255) if is_fg else (255, 255, 255, 255)
    if render_mode == GlyphRenderMode.WHITE_ON_BLACK:
        return (255, 255, 255, 255) if is_fg else (0, 0, 0, 255)
    raise ValueError(f"unsupported render_mode: {render_mode}")


def export_font_glyph_images(
    group_id: str,
    background_path: str,
    image_size: tuple[int, int],
    glyphs: Iterable[FontGlyph],
    output_dir: str,
    file_prefix: str,
    render_mode: GlyphRenderModeLike = GlyphRenderMode.ORIGINAL,
) -> FontGlyphImageExportResult:
    out_dir = Path(output_dir)
    out_dir.mkdir(parents=True, exist_ok=True)
    normalized_mode = normalize_glyph_render_mode(render_mode)

    glyph_list = list(glyphs)
    exported: List[FontGlyphImageItem] = []
    for glyph in glyph_list:
        rgba_2d = glyph.rgba_2d
        if rgba_2d is None:
            raise ValueError("glyph.rgba_2d is required to export glyph images")
        height = len(rgba_2d)
        width = len(rgba_2d[0]) if height > 0 else 0
        image = Image.new("RGBA", (width, height), (0, 0, 0, 0))
        flat_pixels = [_render_rgba_pixel(pixel, render_mode=normalized_mode) for row in rgba_2d for pixel in row]
        image.putdata(flat_pixels)
        file_name = f"{file_prefix}_glyph_{glyph.rect_index:02d}.png"
        target = out_dir / file_name
        image.save(target)
        exported.append(
            FontGlyphImageItem(
                rect_index=glyph.rect_index,
                bbox=glyph.bbox,
                image_path=str(target),
                width=width,
                height=height,
                pixel_count=glyph.pixel_count,
            )
        )

    return FontGlyphImageExportResult(
        group_id=group_id,
        background_path=background_path,
        image_size=image_size,
        output_dir=str(out_dir),
        glyph_images=exported,
        stats={
            "glyph_count": len(exported),
            "exported_count": len(exported),
            "render_mode": normalized_mode.value,
        },
    )

File: typescript-sdk/src/font-glyph-images.ts
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';
import {
  FontGlyph,
  FontGlyphImageExportResult,
  FontGlyphImageItem,
  GlyphRenderMode,
  GlyphRenderModeLike,
  normalizeGlyphRenderMode,
} from './types';

function renderRgbaPixel(pixel: [number, number, number, number], renderMode: GlyphRenderMode): [number, number, number, number] {
  const [r, g, b, a] = pixel;
  if (renderMode === GlyphRenderMode.ORIGINAL) {
    return [r, g, b, a];
  }
  const isFg = a > 0;
  if (renderMode === GlyphRenderMode.BLACK_ON_TRANSPARENT) {
    return isFg ? [0, 0, 0, 255] : [0, 0, 0, 0];
  }
  if (renderMode === GlyphRenderMode.BLACK_ON_WHITE) {
    return isFg ? [0, 0, 0, 255] : [255, 255, 255, 255];
  }
  if (renderMode === GlyphRenderMode.WHITE_ON_BLACK) {
    return isFg ? [255, 255, 255, 255] : [0, 0, 0, 255];
  }
  throw new Error(`unsupported render_mode: ${renderMode}`);
}

export async function exportFontGlyphImages(
  groupId: string,
  backgroundPath: string,
  imageSize: [number, number],
  glyphs: FontGlyph[],
  outputDir: string,
  filePrefix: string,
  renderMode: GlyphRenderModeLike = GlyphRenderMode.ORIGINAL
): Promise<FontGlyphImageExportResult> {
  const outDir = path.resolve(outputDir);
  await fs.promises.mkdir(outDir, { recursive: true });
  const normalizedMode = normalizeGlyphRenderMode(renderMode);

  const exported: FontGlyphImageItem[] = [];
  for (const glyph of glyphs) {
    const rgba2d = glyph.rgba_2d;
    if (!rgba2d) {
      throw new Error('glyph.rgba_2d is required to export glyph images');
    }
    const height = rgba2d.length;
    const width = height > 0 ? rgba2d[0].length : 0;
    const buffer = Buffer.alloc(width * height * 4, 0);
    let offset = 0;
    for (const row of rgba2d) {
      for (const pixel of row) {
        const [r, g, b, a] = renderRgbaPixel(pixel, normalizedMode);
        buffer[offset] = r;
        buffer[offset + 1] = g;
        buffer[offset + 2] = b;
        buffer[offset + 3] = a;
        offset += 4;
      }
    }

    const fileName = `${filePrefix}_glyph_${String(glyph.rect_index).padStart(2, '0')}.png`;
    const target = path.join(outDir, fileName);
    await sharp(buffer, { raw: { width, height, channels: 4 } }).png().toFile(target);

    exported.push({
      rect_index: glyph.rect_index,
      bbox: glyph.bbox,
      image_path: target,
      width,
      height,
      pixel_count: glyph.pixel_count,
    });
  }

  return {
    group_id: groupId,
    background_path: backgroundPath,
    image_size: imageSize,
    output_dir: outDir,
    glyph_images: exported,
    stats: {
      glyph_count: exported.length,
      exported_count: exported.length,
    },
  };
}
